gear-contract-libraries/multitoken/state.rs

```diff
@@ -0,0 +1,124 @@
+            .unwrap_or(&0)
+    }
+
+    fn set_balance(&mut self, account: &ActorId, id: &TokenId, amount: u128) {
+        let mut _balance = self
+            .get_mut()
+            .balances
+            .entry(*id)
+            .or_default()
+            .insert(*account, amount);
+    }
+}
+
+#[derive(Debug, Decode, Encode, TypeInfo)]
+pub enum MTKQuery {
+    Name,
+    Symbol,
+    Uri,
+    BalanceOf(ActorId, TokenId),
+    MetadataOf(TokenId),
+    URI(TokenId),
+    TokensForOwner(ActorId),
+    Supply(TokenId),
+}
+
+#[derive(Debug, Decode, Encode, TypeInfo)]
+pub enum MTKQueryReply {
+    Name(String),
+    Symbol(String),
+    Uri(String),
+    Balance(TokenId),
+    URI(String),
+    MetadataOf(TokenMetadata),
+    TokensForOwner(Vec<TokenId>),
+    Supply(u128),
+}
+
+pub trait MTKTokenState: StateKeeper + BalanceTrait {
+    fn get_uri(&self, id: TokenId) -> String {
+        self.get()
+            .base_uri
+            .clone()
+            .replace("{id}", &format!("{}", id))
+    }
+
+    fn get_metadata(&self, id: TokenId) -> TokenMetadata {
+        self.get()
+            .token_metadata
+            .get(&id)
+            .unwrap_or(&TokenMetadata {
+                ..Default::default()
+            })
+            .clone()
+    }
+
+    fn tokens_for_owner(&self, owner: &ActorId) -> Vec<TokenId> {
+        let mut tokens: Vec<TokenId> = Vec::new();
+        let balances = &self.get().balances;
+        for (token, bals) in balances {
+            if let Some(_user) = bals.get(owner) {
+                tokens.push(*token);
+            }
+        }
+        tokens
+    }
+
+    fn supply(&self, id: TokenId) -> u128 {
+        self.get()
+            .balances
+            .clone()
+            .entry(id)
+            .or_default()
+            .clone()
+            .into_values()
+            .collect::<Vec<u128>>()
+            .iter()
+            .sum()
+    }
+
+    fn proc_state(&mut self, query: MTKQuery) -> Option<Vec<u8>> {
+        let encoded = match query {
+            MTKQuery::Name => MTKQueryReply::Name(self.get().name.clone()).encode(),
+            MTKQuery::Symbol => MTKQueryReply::Symbol(self.get().symbol.clone()).encode(),
+            MTKQuery::Uri => MTKQueryReply::Uri(self.get().base_uri.clone()).encode(),
+            MTKQuery::BalanceOf(account, id) => {
+                MTKQueryReply::Balance(Self::get_balance(self, &account, &id)).encode()
+            }
+            MTKQuery::URI(id) => MTKQueryReply::URI(Self::get_uri(self, id)).encode(),
+            MTKQuery::MetadataOf(id) => {
+                MTKQueryReply::MetadataOf(Self::get_metadata(self, id)).encode()
+            }
+            MTKQuery::TokensForOwner(owner) => {
+                MTKQueryReply::TokensForOwner(Self::tokens_for_owner(self, &owner)).encode()
+            }
+            MTKQuery::Supply(id) => MTKQueryReply::Supply(Self::supply(self, id)).encode(),
+        };
+        Some(encoded)
```

Comment:
No need to repeat `encode` in every line.

```suggestion
        let state = match query {
            MTKQuery::Name => MTKQueryReply::Name(self.get().name.clone()),
            MTKQuery::Symbol => MTKQueryReply::Symbol(self.get().symbol.clone()),
            MTKQuery::Uri => MTKQueryReply::Uri(self.get().base_uri.clone()),
            MTKQuery::BalanceOf(account, id) => {
                MTKQueryReply::Balance(Self::get_balance(self, &account, &id))
            }
            MTKQuery::URI(id) => MTKQueryReply::URI(Self::get_uri(self, id)),
            MTKQuery::MetadataOf(id) => {
                MTKQueryReply::MetadataOf(Self::get_metadata(self, id))
            }
            MTKQuery::TokensForOwner(owner) => {
                MTKQueryReply::TokensForOwner(Self::tokens_for_owner(self, &owner))
            }
            MTKQuery::Supply(id) => MTKQueryReply::Supply(Self::supply(self, id)),
        };
        Some(state.encode())
```